ccl/sqlproxyccl/acl: TestParsingErrorHandling failed

ccl/sqlproxyccl/acl.TestParsingErrorHandling [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9113225?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9113225?buildTab=artifacts#/) on master @ [11fbbca584e3bd774ebee3ee94c49ca30a096b26](https://github.com/cockroachdb/cockroach/commits/11fbbca584e3bd774ebee3ee94c49ca30a096b26):


```
=== RUN   TestParsingErrorHandling
--- FAIL: TestParsingErrorHandling (0.82s)
=== RUN   TestParsingErrorHandling/error_after_update
E230317 06:40:49.045938 115 ccl/sqlproxyccl/acl/file.go:71  [-] 13  error updating access control list from file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling3321990234/001/error_file.idk modified at 2023-03-17 06:40:49.042765008 +0000 UTC: yaml: unmarshal errors:
E230317 06:40:49.045938 115 ccl/sqlproxyccl/acl/file.go:71  [-] 13 +  line 1: cannot unmarshal !!str ``not yaml`` into acl.AllowlistFile
    file_test.go:499: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/acl/file_test.go:499
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestParsingErrorHandling/error_after_update
E230317 06:40:49.863088 117 ccl/sqlproxyccl/acl/file.go:71  [-] 14  error updating access control list from file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling3321990234/001/allowlist.yaml modified at 2023-03-17 06:40:49.042765008 +0000 UTC: yaml: unmarshal errors:
E230317 06:40:49.863088 117 ccl/sqlproxyccl/acl/file.go:71  [-] 14 +  line 1: cannot unmarshal !!str ``no long...`` into acl.AllowlistFile
E230317 06:40:49.863232 117 ccl/sqlproxyccl/acl/file.go:110  [-] 15  Could not read file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling3321990234/001/allowlist.yaml for *acl.Allowlist: yaml: unmarshal errors:
E230317 06:40:49.863232 117 ccl/sqlproxyccl/acl/file.go:110  [-] 15 +  line 1: cannot unmarshal !!str ``no long...`` into acl.AllowlistFile
E230317 06:40:49.863289 117 ccl/sqlproxyccl/acl/file.go:100  [-] 16  WatchList daemon stopped: context canceled
    --- FAIL: TestParsingErrorHandling/error_after_update (0.20s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-sessions @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParsingErrorHandling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25540

ccl/sqlproxyccl/acl.TestParsingErrorHandling [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9113225?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9113225?buildTab=artifacts#/) on master @ [11fbbca584e3bd774ebee3ee94c49ca30a096b26](https://github.com/cockroachdb/cockroach/commits/11fbbca584e3bd774ebee3ee94c49ca30a096b26):


```
=== RUN   TestParsingErrorHandling
--- FAIL: TestParsingErrorHandling (0.43s)
=== RUN   TestParsingErrorHandling/error_after_update
E230317 06:41:09.756412 62 ccl/sqlproxyccl/acl/file.go:71  [-] 13  error updating access control list from file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling4216695855/001/error_file.idk modified at 2023-03-17 06:41:09.750478078 +0000 UTC: yaml: unmarshal errors:
E230317 06:41:09.756412 62 ccl/sqlproxyccl/acl/file.go:71  [-] 13 +  line 1: cannot unmarshal !!str ``not yaml`` into acl.AllowlistFile
    file_test.go:499: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/acl/file_test.go:499
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestParsingErrorHandling/error_after_update
E230317 06:41:10.185336 64 ccl/sqlproxyccl/acl/file.go:71  [-] 14  error updating access control list from file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling4216695855/001/allowlist.yaml modified at 2023-03-17 06:41:09.750478078 +0000 UTC: yaml: unmarshal errors:
E230317 06:41:10.185336 64 ccl/sqlproxyccl/acl/file.go:71  [-] 14 +  line 1: cannot unmarshal !!str ``no long...`` into acl.AllowlistFile
E230317 06:41:10.185583 64 ccl/sqlproxyccl/acl/file.go:110  [-] 15  Could not read file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling4216695855/001/allowlist.yaml for *acl.Allowlist: yaml: unmarshal errors:
E230317 06:41:10.185583 64 ccl/sqlproxyccl/acl/file.go:110  [-] 15 +  line 1: cannot unmarshal !!str ``no long...`` into acl.AllowlistFile
E230317 06:41:10.185642 64 ccl/sqlproxyccl/acl/file.go:100  [-] 16  WatchList daemon stopped: context canceled
    --- FAIL: TestParsingErrorHandling/error_after_update (0.20s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParsingErrorHandling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/acl.TestParsingErrorHandling [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9202044?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9202044?buildTab=artifacts#/) on master @ [a930643519faded2232f84850b30b12e7a5a1302](https://github.com/cockroachdb/cockroach/commits/a930643519faded2232f84850b30b12e7a5a1302):


```
=== RUN   TestParsingErrorHandling
--- FAIL: TestParsingErrorHandling (0.85s)
=== RUN   TestParsingErrorHandling/error_after_update
    file_test.go:499: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/acl/file_test.go:499
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestParsingErrorHandling/error_after_update
E230322 23:08:20.241296 98 ccl/sqlproxyccl/acl/file.go:71  [-] 14  error updating access control list from file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling2544467519/001/allowlist.yaml modified at 2023-03-22 23:08:19.392641798 +0000 UTC: yaml: unmarshal errors:
E230322 23:08:20.241296 98 ccl/sqlproxyccl/acl/file.go:71  [-] 14 +  line 1: cannot unmarshal !!str ``no long...`` into acl.AllowlistFile
E230322 23:08:20.241427 98 ccl/sqlproxyccl/acl/file.go:110  [-] 15  Could not read file /artifacts/tmp/_tmp/72c09b96526627cf742479f65b1af982/TestParsingErrorHandling2544467519/001/allowlist.yaml for *acl.Allowlist: yaml: unmarshal errors:
E230322 23:08:20.241427 98 ccl/sqlproxyccl/acl/file.go:110  [-] 15 +  line 1: cannot unmarshal !!str ``no long...`` into acl.AllowlistFile
E230322 23:08:20.241485 98 ccl/sqlproxyccl/acl/file.go:100  [-] 16  WatchList daemon stopped: context canceled
    --- FAIL: TestParsingErrorHandling/error_after_update (0.50s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParsingErrorHandling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/sqlproxyccl/acl/BUILD.bazel
+++ b/pkg/ccl/sqlproxyccl/acl/BUILD.bazel
@@ -33,6 +33,7 @@ go_test(
     embed = [":acl"],
     tags = ["ccl_test"],
     deps = [
+        "//pkg/testutils",
         "//pkg/util/leaktest",
         "//pkg/util/metric",
         "//pkg/util/timeutil",

--- a/pkg/ccl/sqlproxyccl/acl/file_test.go
+++ b/pkg/ccl/sqlproxyccl/acl/file_test.go
@@ -11,6 +11,7 @@ package acl
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"net"
 	"os"
@@ -18,6 +19,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -491,30 +493,43 @@ func TestParsingErrorHandling(t *testing.T) {
 
 		// Update with invalid data
 		require.NoError(t, os.WriteFile(filename, []byte(`no longer valid yaml`), 0777))
-		select {
-		case <-next:
-			t.Error("should not have gotten a new controller")
-		case <-time.After(time.Millisecond * 500):
-			// error count should go up
-			require.Equal(t, int64(1), errorCountMetric.Snapshot().Value())
-		}
+
+		testutils.SucceedsSoon(t, func() error {
+			select {
+			case <-next:
+				t.Fatal("should not have gotten a new controller")
+				// We need to return something to make the compiler happy, but t.Fatal will end execution.
+				return nil
+			default:
+				errorCount := errorCountMetric.Snapshot().Value()
+				// If error count isn't one, then it hasn't happened yet.
+				if errorCount != 1 {
+					return fmt.Errorf("Expected error count to be 1 but got %d", errorCount)
+				}
+
+				return nil
+			}
+		})
 
 		// Update with valid data now
 		require.NoError(t, os.WriteFile(filename, []byte(`{"allowlist":{"tenant":{"ips": ["2.2.2.2/32"]}}}`), 0777))
-		select {
-		case controller := <-next:
-			// error count should go down
-			require.Equal(t, int64(0), errorCountMetric.Snapshot().Value())
-			allowlist := controller.(*Allowlist)
-			require.Equal(t, map[string]AllowEntry{
-				"tenant": {
-					ips: []*net.IPNet{
-						parseIPNet("2.2.2.2/32"),
+		testutils.SucceedsSoon(t, func() error {
+			select {
+			case controller := <-next:
+				// error count should go down
+				require.Equal(t, int64(0), errorCountMetric.Snapshot().Value())
+				allowlist := controller.(*Allowlist)
+				require.Equal(t, map[string]AllowEntry{
+					"tenant": {
+						ips: []*net.IPNet{
+							parseIPNet("2.2.2.2/32"),
+						},
 					},
-				},
-			}, allowlist.entries)
-		case <-time.After(time.Millisecond * 500):
-			t.Error("should have gotten a new controller")
-		}
+				}, allowlist.entries)
+				return nil
+			default:
+				return errors.New("new controller not created yet")
+			}
+		})
 	})
 }

